File: examples/basic-host-swift/Sources/BasicHostApp/McpHostViewModel.swift
import Foundation
import SwiftUI
import MCP
import McpApps
import os.log

private let logger = Logger(subsystem: "com.example.BasicHostSwift", category: "ToolCall")

/// View model managing MCP server connection and tool execution.
@MainActor
class McpHostViewModel: ObservableObject {
    // MARK: - Published State

    @Published var connectionState: ConnectionState = .disconnected
    @Published var tools: [Tool] = []
    @Published var toolInputJson: String = "{}"

    /// Selected tool - updates input JSON with defaults when changed
    @Published var selectedTool: Tool? {
        didSet {
            if let tool = selectedTool {
                toolInputJson = generateDefaultInput(for: tool)
            }
        }
    }
    @Published var activeToolCalls: [ToolCallInfo] = []
    @Published var errorMessage: String?
    @Published var toastMessage: String?

    func showToast(_ message: String, duration: TimeInterval = 3.0) {
        toastMessage = message
        Task {
            try? await Task.sleep(nanoseconds: UInt64(duration * 1_000_000_000))
            await MainActor.run { self.toastMessage = nil }
        }
    }

    /// Discovered MCP servers (name, url pairs)
    @Published var discoveredServers: [(name: String, url: String)] = []

    /// Whether server discovery is in progress
    @Published var isDiscovering = false

    /// Discovery configuration
    static let basePort = 3101
    static let discoveryTimeout: TimeInterval = 1.0  // 1 second
    static let baseHost = "localhost"

    /// Selected server index (-1 for custom URL)
    @Published var selectedServerIndex: Int = 0

    /// Custom server URL (used when selectedServerIndex is -1)
    @Published var customServerUrl: String = ""

    /// Computed server URL based on selection
    var serverUrlString: String {
        if selectedServerIndex >= 0 && selectedServerIndex < discoveredServers.count {
            return discoveredServers[selectedServerIndex].url
        }
        return customServerUrl
    }

    // MARK: - Private State

    private var mcpClient: Client?

    private let hostInfo = Implementation(name: "BasicHostSwift", version: "1.0.0")

    private let hostCapabilities = McpUiHostCapabilities(
        openLinks: EmptyCapability(),
        serverTools: ServerToolsCapability(),
        serverResources: ServerResourcesCapability(),
        logging: EmptyCapability()
    )

    init() {}

    // MARK: - Server Discovery

    /// Discover available MCP servers starting from basePort
    func discoverServers() async {
        isDiscovering = true
        discoveredServers = []
        var port = Self.basePort

        while true {
            let url = "http://\(Self.baseHost):\(port)/mcp"

            if let serverName = await tryConnect(url: url, timeout: Self.discoveryTimeout) {
                discoveredServers.append((name: serverName, url: url))
                port += 1
            } else {
                // Connection failed or timed out - stop discovery
                break
            }
        }

        isDiscovering = false

        // Auto-connect to first discovered server
        if !discoveredServers.isEmpty {
            selectedServerIndex = 0
            await connect()
        }
    }

    /// Try to connect to a server URL with timeout, returns server name on success
    private func tryConnect(url: String, timeout: TimeInterval) async -> String? {
        guard let serverUrl = URL(string: url) else { return nil }

        do {
            let result = try await withThrowingTaskGroup(of: String.self) { group in
                group.addTask {
                    let client = Client(name: self.hostInfo.name, version: self.hostInfo.version)
                    let transport = HTTPClientTransport(endpoint: serverUrl)
                    let initResult = try await client.connect(transport: transport)
                    let name = initResult.serverInfo.name
                    await client.disconnect()
                    return name
                }

                group.addTask {
                    try await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                    throw CancellationError()
                }

                guard let result = try await group.next() else {
                    throw CancellationError()
                }
                group.cancelAll()
                return result
            }
            return result
        } catch {
            return nil
        }
    }

    // MARK: - Connection Management

    func connect() async {
        connectionState = .connecting
        errorMessage = nil

        do {
            guard let serverUrl = URL(string: serverUrlString) else {
                throw ConnectionError.invalidUrl(serverUrlString)
            }

            let client = Client(name: hostInfo.name, version: hostInfo.version)
            let transport = HTTPClientTransport(endpoint: serverUrl)
            _ = try await client.connect(transport: transport)

            self.mcpClient = client

            let (toolsList, _) = try await client.listTools()
            self.tools = toolsList

            connectionState = .connected

            if !tools.isEmpty {
                selectedTool = tools[0]
            }
        } catch {
            connectionState = .error(error.localizedDescription)
            errorMessage = "Failed to connect: \(error.localizedDescription)"
        }
    }

    func disconnect() async {
        await mcpClient?.disconnect()
        mcpClient = nil
        tools = []
        selectedTool = nil
        connectionState = .disconnected
    }

    /// Switch to a different server (disconnect current, connect to new)
    func switchServer(to index: Int) async {
        // Don't switch if already on this server and connected
        if index == selectedServerIndex && connectionState == .connected {
            return
        }

        // Disconnect from current server if connected
        if connectionState == .connected {
            await disconnect()
        }

        // Update selection and connect
        selectedServerIndex = index
        await connect()
    }

    // MARK: - Tool Execution

    func callTool() async {
        guard let tool = selectedTool else { return }
        guard let client = mcpClient else { return }

        errorMessage = nil

        do {
            guard let inputData = toolInputJson.data(using: .utf8),
                  let inputDict = try JSONSerialization.jsonObject(with: inputData) as? [String: Any] else {
                throw ToolCallError.invalidJson
            }

            let serverName = selectedServerIndex >= 0 && selectedServerIndex < discoveredServers.count
                ? discoveredServers[selectedServerIndex].name
                : "Custom Server"

            let toolCallInfo = ToolCallInfo(
                serverName: serverName,
                tool: tool,
                input: inputDict,
                client: client,
                hostInfo: hostInfo,
                hostCapabilities: hostCapabilities
            )

            activeToolCalls.append(toolCallInfo)

            Task {
                await toolCallInfo.execute()
            }

        } catch {
            errorMessage = "Failed to call tool: \(error.localizedDescription)"
        }
    }

    /// Generate default input JSON from tool's inputSchema
    private func generateDefaultInput(for tool: Tool) -> String {
        var defaults: [String: Any] = [:]

        // inputSchema is typically: { "type": "object", "properties": { ... }, "required": [...] }
        guard case .object(let schema) = tool.inputSchema,
              case .object(let properties)? = schema["properties"] else {
            return "{}"
        }

        // Get required fields
        var requiredFields = Set<String>()
        if case .array(let required)? = schema["required"] {
            for item in required {
                if case .string(let field) = item {
                    requiredFields.insert(field)
                }
            }
        }

        for (key, propSchema) in properties {
            guard case .object(let prop) = propSchema else { continue }

            // Check for explicit default value
            if let defaultValue = prop["default"] {
                defaults[key] = valueToAny(defaultValue)
                continue
            }

            // For required fields or all fields, generate appropriate defaults
            if let typeValue = prop["type"] {
                if case .string(let type) = typeValue {
                    switch type {
                    case "string":
                        // Check for enum values
                        if case .array(let enumValues)? = prop["enum"], let first = enumValues.first {
                            defaults[key] = valueToAny(first)
                        } else {
                            defaults[key] = ""
                        }
                    case "number":
                        defaults[key] = 0.0
                    case "integer":
                        defaults[key] = 0
                    case "boolean":
                        defaults[key] = false
                    case "array":
                        defaults[key] = []
                    case "object":
                        defaults[key] = [:]
                    default:
                        if requiredFields.contains(key) {
                            defaults[key] = NSNull()
                        }
                    }
                }
            }
        }

        // Convert to JSON string
        if let data = try? JSONSerialization.data(withJSONObject: defaults, options: [.prettyPrinted, .sortedKeys]),
           let json = String(data: data, encoding: .utf8) {
            return json
        }
        return "{}"
    }

    /// Convert MCP Value to Any for JSON serialization
    private func valueToAny(_ value: Value) -> Any {
        switch value {
        case .string(let s): return s
        case .int(let i): return i
        case .double(let d): return d
        case .bool(let b): return b
        case .array(let arr): return arr.map { valueToAny($0) }
        case .object(let obj): return obj.mapValues { valueToAny($0) }
        case .null: return NSNull()
        case .data(_, let d): return d.base64EncodedString()
        }
    }

    func removeToolCall(_ toolCall: ToolCallInfo) async {
        if let error = await toolCall.teardown() {
            showToast(error)
        }
        activeToolCalls.removeAll { $0.id == toolCall.id }
    }
}

// MARK: - Connection State

enum ConnectionState: Equatable {
    case disconnected
    case connecting
    case connected
    case error(String)

    var description: String {
        switch self {
        case .disconnected: return "Disconnected"
        case .connecting: return "Connecting..."
        case .connected: return "Connected"
        case .error(let message): return "Error: \(message)"
        }
    }
}

// MARK: - Tool Call Info

/// Tool result type matching MCP SDK's callTool return
struct ToolResult {
    let content: [Tool.Content]
    let isError: Bool?
}

@MainActor
class ToolCallInfo: ObservableObject, Identifiable {
    let id = UUID()
    let serverName: String
    let tool: Tool
    let input: [String: Any]
    let client: Client
    let hostInfo: Implementation
    let hostCapabilities: McpUiHostCapabilities

    @Published var appBridge: AppBridge?
    @Published var htmlContent: String?
    @Published var cspConfig: CspConfig?
    @Published var preferredHeight: CGFloat = 350
    @Published var result: ToolResult?
    @Published var state: ExecutionState = .calling
    @Published var error: String?
    @Published var isTearingDown = false

    init(
        serverName: String,
        tool: Tool,
        input: [String: Any],
        client: Client,
        hostInfo: Implementation,
        hostCapabilities: McpUiHostCapabilities
    ) {
        self.serverName = serverName
        self.tool = tool
        self.input = input
        self.client = client
        self.hostInfo = hostInfo
        self.hostCapabilities = hostCapabilities
    }

    func execute() async {
        do {
            state = .calling

            // Convert input to MCP Value type
            let arguments = input.compactMapValues { value -> Value? in
                if let str = value as? String { return .string(str) }
                if let num = value as? Int { return .int(num) }
                if let num = value as? Double { return .double(num) }
                if let bool = value as? Bool { return .bool(bool) }
                return nil
            }

            let (content, isError) = try await client.callTool(name: tool.name, arguments: arguments)
            self.result = ToolResult(content: content, isError: isError)

            if let uiResourceUri = getUiResourceUri(from: tool) {
                state = .loadingUi
                try await loadUiResource(uri: uiResourceUri)
                state = .ready
            } else {
                state = .completed
            }
        } catch {
            state = .error
            self.error = error.localizedDescription
        }
    }

    private func getUiResourceUri(from tool: Tool) -> String? {
        // Access the _meta field to get the UI resource URI (requires swift-sdk with _meta support)
        if let meta = tool._meta,
           let uriValue = meta[McpAppsConfig.resourceUriMetaKey],
           case .string(let uri) = uriValue {
            return uri
        }
        return nil
    }

    private func loadUiResource(uri: String) async throws {
        let contents = try await client.readResource(uri: uri)

        guard let content = contents.first else {
            throw ToolCallError.noResourceContent
        }

        guard content.mimeType == McpAppsConfig.resourceMimeType else {
            throw ToolCallError.invalidMimeType(content.mimeType ?? "unknown")
        }

        if let text = content.text {
            htmlContent = text
        } else if let blob = content.blob,
                  let data = Data(base64Encoded: blob),
                  let text = String(data: data, encoding: .utf8) {
            htmlContent = text
        } else {
            throw ToolCallError.invalidHtmlContent
        }
    }

    func setupAppBridge(transport: WKWebViewTransport) async throws {
        let bridge = AppBridge(
            hostInfo: hostInfo,
            hostCapabilities: hostCapabilities,
            options: HostOptions(
                hostContext: McpUiHostContext(
                    theme: .light,
                    displayMode: .inline,
                    platform: .mobile,
                    deviceCapabilities: DeviceCapabilities(touch: true, hover: false)
                )
            )
        )

        bridge.onInitialized = { [weak self] in
            Task { @MainActor in
                guard let self = self else { return }
                let params = McpUiToolInputParams(
                    arguments: self.input.mapValues { AnyCodable($0) }
                )
                try? await bridge.sendToolInput(params)
                if let result = self.result {
                    try? await self.sendToolResult(result, to: bridge)
                }
            }
        }

        bridge.onMessage = { params in
            print("[Host] Message from Guest UI: \(params.role)")
            return McpUiMessageResult(isError: false)
        }

        bridge.onOpenLink = { params in
            print("[Host] Open link request: \(params.url)")
            if let urlObj = URL(string: params.url) {
                await MainActor.run {
                    UIApplication.shared.open(urlObj)
                }
            }
            return McpUiOpenLinkResult(isError: false)
        }

        bridge.onLoggingMessage = { params in
            print("[Host] Guest UI log [\(params.level)]: \(params.data.value)")
        }

        bridge.onSizeChange = { [weak self] params in
            print("[Host] Size change: \(params.width ?? 0) x \(params.height ?? 0)")
            if let height = params.height {
                Task { @MainActor in
                    self?.preferredHeight = CGFloat(height)
                }
            }
        }

        bridge.onToolCall = { [weak self] toolName, arguments in
            guard let self = self else { throw ToolCallError.clientNotAvailable }

            let args = arguments?.compactMapValues { value -> Value? in
                if let str = value.value as? String { return .string(str) }
                if let num = value.value as? Int { return .int(num) }
                if let num = value.value as? Double { return .double(num) }
                if let bool = value.value as? Bool { return .bool(bool) }
                return nil
            }

            let (content, isError) = try await self.client.callTool(name: toolName, arguments: args)

            return [
                "content": AnyCodable(content.map { c -> [String: Any] in
                    switch c {
                    case .text(let text):
                        return ["type": "text", "text": text]
                    case .image(let data, let mimeType, _):
                        return ["type": "image", "data": data, "mimeType": mimeType]
                    default:
                        return ["type": "text", "text": ""]
                    }
                }),
                "isError": AnyCodable(isError ?? false)
            ]
        }

        bridge.onResourceRead = { [weak self] uri in
            guard let self = self else { throw ToolCallError.clientNotAvailable }

            let contents = try await self.client.readResource(uri: uri)

            return [
                "contents": AnyCodable(contents.map { c in
                    var dict: [String: Any] = ["uri": c.uri]
                    if let text = c.text { dict["text"] = text }
                    if let blob = c.blob { dict["blob"] = blob }
                    if let mimeType = c.mimeType { dict["mimeType"] = mimeType }
                    return dict
                })
            ]
        }

        logger.info("Connecting bridge to transport...")
        try await bridge.connect(transport)
        self.appBridge = bridge
        logger.info("AppBridge is now set and ready")
    }

    private func sendToolResult(_ result: ToolResult, to bridge: AppBridge) async throws {
        let contentItems: [McpUiToolResultNotificationParamsContentItem] = result.content.compactMap { c in
            switch c {
            case .text(let text):
                return .text(McpUiToolResultNotificationParamsContentItemText(type: "text", text: text))
            case .image(let data, let mimeType, _):
                return .image(McpUiToolResultNotificationParamsContentItemImage(type: "image", data: data, mimeType: mimeType))
            default:
                return nil
            }
        }
        let params = McpUiToolResultParams(content: contentItems, isError: result.isError)
        try await bridge.sendToolResult(params)
    }

    /// Teardown the app bridge before removing the tool call
    /// Returns an error message if teardown failed, nil on success
    func teardown() async -> String? {
        // Prevent double-tap
        guard !isTearingDown else {
            logger.info("Teardown already in progress, skipping")
            return nil
        }
        isTearingDown = true
        logger.info("Starting teardown, appBridge exists: \(self.appBridge != nil)")

        var errorMessage: String?

        if let bridge = appBridge {
            // Only send teardown if the bridge is initialized
            // If not initialized, the app hasn't done anything worth saving
            let isReady = await bridge.isReady()
            if isReady {
                logger.info("Sending teardown request...")
                do {
                    _ = try await bridge.sendResourceTeardown()
                    logger.info("Teardown request completed successfully")
                } catch {
                    logger.error("Teardown request failed: \(String(describing: error))")
                    errorMessage = "Teardown failed: app may not have saved data"
                }
            } else {
                logger.info("Skipping teardown - bridge not yet initialized")
            }

            logger.info("Closing bridge...")
            await bridge.close()
            logger.info("Bridge closed")
        } else {
            logger.warning("No bridge to teardown (appBridge is nil)")
        }
        appBridge = nil
        logger.info("Teardown complete, will remove card from list")
        return errorMessage
    }
}

// MARK: - Execution State

enum ExecutionState {
    case calling, loadingUi, ready, completed, error

    var description: String {
        switch self {
        case .calling: return "Calling..."
        case .loadingUi: return "Loading UI..."
        case .ready: return "Ready"
        case .completed: return "Completed"
        case .error: return "Error"
        }
    }
}

// MARK: - Errors

enum ToolCallError: LocalizedError {
    case invalidJson
    case noResourceContent
    case invalidMimeType(String)
    case invalidHtmlContent
    case clientNotAvailable

    var errorDescription: String? {
        switch self {
        case .invalidJson: return "Invalid JSON input"
        case .noResourceContent: return "No content in UI resource"
        case .invalidMimeType(let m): return "Invalid MIME type: \(m)"
        case .invalidHtmlContent: return "Invalid HTML content"
        case .clientNotAvailable: return "MCP client not available"
        }
    }
}

enum ConnectionError: LocalizedError {
    case invalidUrl(String)

    var errorDescription: String? {
        switch self {
        case .invalidUrl(let url): return "Invalid URL: \(url)"
        }
    }
}

File: examples/basic-host-kotlin/src/main/kotlin/com/example/mcpappshost/McpHostViewModel.kt
package com.example.mcpappshost

import android.util.Log
import android.webkit.WebView
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.modelcontextprotocol.apps.generated.*
import io.modelcontextprotocol.kotlin.sdk.client.Client
import io.modelcontextprotocol.kotlin.sdk.client.SseClientTransport
import io.modelcontextprotocol.kotlin.sdk.types.ReadResourceRequest
import io.modelcontextprotocol.kotlin.sdk.types.ReadResourceRequestParams
import io.modelcontextprotocol.kotlin.sdk.types.TextResourceContents
import io.modelcontextprotocol.kotlin.sdk.types.BlobResourceContents
import io.modelcontextprotocol.kotlin.sdk.types.TextContent
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.sse.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.*
import kotlinx.serialization.encodeToString

private const val TAG = "McpHostViewModel"

data class DiscoveredServer(val name: String, val url: String)

data class ToolInfo(
    val name: String,
    val description: String?,
    val inputSchema: JsonElement?,
    val uiResourceUri: String? = null  // From _meta["ui/resourceUri"]
)

sealed class ConnectionState {
    data object Disconnected : ConnectionState()
    data object Connecting : ConnectionState()
    data object Connected : ConnectionState()
    data class Error(val message: String) : ConnectionState()
}

data class ToolCallState(
    val id: String = java.util.UUID.randomUUID().toString(),
    val serverName: String,
    val toolName: String,
    val input: String,
    val inputArgs: Map<String, Any>? = null,
    val state: State = State.CALLING,
    val result: String? = null,
    val toolResult: String? = null,  // Raw tool result for AppBridge
    val error: String? = null,
    val htmlContent: String? = null,
    var webView: WebView? = null,
    var preferredHeight: Int = 350,
    var appBridgeConnected: Boolean = false,
    val isDestroying: Boolean = false  // Two-phase teardown: true while waiting for app response
) {
    enum class State { CALLING, LOADING_UI, READY, COMPLETED, ERROR }

    val hasApp: Boolean get() = htmlContent != null && state == State.READY
}

class McpHostViewModel : ViewModel() {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    companion object {
        const val BASE_PORT = 3101
        const val DISCOVERY_TIMEOUT_MS = 1000L
        // Android emulator uses 10.0.2.2 for host machine's localhost
        const val BASE_HOST = "10.0.2.2"
    }

    // Connection state
    private val _connectionState = MutableStateFlow<ConnectionState>(ConnectionState.Disconnected)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    // Server discovery
    private val _discoveredServers = MutableStateFlow<List<DiscoveredServer>>(emptyList())
    val discoveredServers: StateFlow<List<DiscoveredServer>> = _discoveredServers.asStateFlow()

    private val _isDiscovering = MutableStateFlow(false)
    val isDiscovering: StateFlow<Boolean> = _isDiscovering.asStateFlow()

    // Tools
    private val _tools = MutableStateFlow<List<ToolInfo>>(emptyList())
    val tools: StateFlow<List<ToolInfo>> = _tools.asStateFlow()

    private val _selectedTool = MutableStateFlow<ToolInfo?>(null)
    val selectedTool: StateFlow<ToolInfo?> = _selectedTool.asStateFlow()

    // Server selection
    private val _selectedServerIndex = MutableStateFlow(0)
    val selectedServerIndex: StateFlow<Int> = _selectedServerIndex.asStateFlow()

    // Tool input
    private val _toolInputJson = MutableStateFlow("{}")
    val toolInputJson: StateFlow<String> = _toolInputJson.asStateFlow()

    // Active tool calls
    private val _toolCalls = MutableStateFlow<List<ToolCallState>>(emptyList())
    val toolCalls: StateFlow<List<ToolCallState>> = _toolCalls.asStateFlow()

    private var mcpClient: Client? = null

    private val hostInfo = Implementation(name = "BasicHostKotlin", version = "1.0.0")
    private val hostCapabilities = McpUiHostCapabilities(
        openLinks = EmptyCapability,
        serverTools = McpUiHostCapabilitiesServerTools(),
        serverResources = McpUiHostCapabilitiesServerResources(),
        logging = EmptyCapability
    )

    init {
        // Start discovery on launch
        discoverServers()
    }

    fun discoverServers() {
        viewModelScope.launch {
            _isDiscovering.value = true
            _discoveredServers.value = emptyList()

            val discovered = mutableListOf<DiscoveredServer>()
            var port = BASE_PORT

            while (true) {
                val url = "http://$BASE_HOST:$port/sse"
                val serverName = tryConnect(url)

                if (serverName != null) {
                    discovered.add(DiscoveredServer(serverName, url))
                    _discoveredServers.value = discovered.toList()
                    Log.i(TAG, "Discovered server: $serverName at $url")
                    port++
                } else {
                    Log.i(TAG, "No server at port $port, stopping discovery")
                    break
                }
            }

            _isDiscovering.value = false
            Log.i(TAG, "Discovery complete, found ${discovered.size} servers")

            // Auto-connect to first discovered server
            if (discovered.isNotEmpty()) {
                _selectedServerIndex.value = 0
                connect()
            }
        }
    }

    private suspend fun tryConnect(url: String): String? {
        return try {
            withTimeout(DISCOVERY_TIMEOUT_MS) {
                val httpClient = HttpClient(CIO) {
                    install(SSE)
                }
                try {
                    val transport = SseClientTransport(httpClient, url)
                    val client = Client(
                        clientInfo = io.modelcontextprotocol.kotlin.sdk.types.Implementation(
                            name = "BasicHostKotlin",
                            version = "1.0.0"
                        )
                    )
                    client.connect(transport)
                    val serverName = client.serverVersion?.name ?: url
                    serverName
                } finally {
                    httpClient.close()
                }
            }
        } catch (e: Exception) {
            Log.d(TAG, "Discovery failed for $url: ${e.message}")
            null
        }
    }

    fun selectTool(tool: ToolInfo) {
        _selectedTool.value = tool
        // Generate default input from schema
        _toolInputJson.value = generateDefaultInput(tool)
    }

    fun updateToolInput(input: String) {
        _toolInputJson.value = input
    }

    fun switchServer(index: Int) {
        if (index == _selectedServerIndex.value && _connectionState.value is ConnectionState.Connected) {
            return
        }
        viewModelScope.launch {
            disconnect()
            _selectedServerIndex.value = index
            connect()
        }
    }

    fun connect() {
        val servers = _discoveredServers.value
        val serverUrl = if (_selectedServerIndex.value >= 0 && _selectedServerIndex.value < servers.size) {
            servers[_selectedServerIndex.value].url
        } else {
            return
        }

        viewModelScope.launch {
            try {
                _connectionState.value = ConnectionState.Connecting
                Log.i(TAG, "Connecting to $serverUrl")

                val httpClient = HttpClient(CIO) {
                    install(SSE)
                }
                val transport = SseClientTransport(httpClient, serverUrl)

                val client = Client(
                    clientInfo = io.modelcontextprotocol.kotlin.sdk.types.Implementation(
                        name = "BasicHostKotlin",
                        version = "1.0.0"
                    )
                )
                client.connect(transport)

                mcpClient = client

                // List tools
                val result = client.listTools()
                _tools.value = result.tools.map { tool ->
                    // Extract UI resource URI from _meta (JsonObject)
                    val meta = tool.meta as? JsonObject
                    val uiResourceUri = meta?.get("ui/resourceUri")?.let { element ->
                        (element as? JsonPrimitive)?.contentOrNull
                    }
                    Log.d(TAG, "Tool ${tool.name} uiResourceUri: $uiResourceUri")
                    ToolInfo(
                        name = tool.name,
                        description = tool.description,
                        inputSchema = null,
                        uiResourceUri = uiResourceUri
                    )
                }

                if (_tools.value.isNotEmpty()) {
                    selectTool(_tools.value.first())
                }

                _connectionState.value = ConnectionState.Connected
                Log.i(TAG, "Connected, found ${_tools.value.size} tools")

            } catch (e: Exception) {
                Log.e(TAG, "Connection failed", e)
                _connectionState.value = ConnectionState.Error(e.message ?: "Unknown error")
            }
        }
    }

    fun disconnect() {
        mcpClient = null
        _tools.value = emptyList()
        _selectedTool.value = null
        _connectionState.value = ConnectionState.Disconnected
    }

    fun callTool() {
        val tool = _selectedTool.value ?: return
        val client = mcpClient ?: return

        val servers = _discoveredServers.value
        val serverName = if (_selectedServerIndex.value in servers.indices) {
            servers[_selectedServerIndex.value].name
        } else "Custom"

        val toolCall = ToolCallState(
            serverName = serverName,
            toolName = tool.name,
            input = _toolInputJson.value
        )
        _toolCalls.value = _toolCalls.value + toolCall

        viewModelScope.launch {
            try {
                // Parse input JSON
                val inputArgs = try {
                    json.parseToJsonElement(_toolInputJson.value) as? JsonObject
                } catch (e: Exception) {
                    null
                }

                // Call the tool (name, arguments, meta, options)
                val callResult = client.callTool(tool.name, emptyMap(), emptyMap())

                // Check for UI resource
                if (tool.uiResourceUri != null) {
                    updateToolCall(toolCall.id) { it.copy(state = ToolCallState.State.LOADING_UI) }
                    Log.i(TAG, "Reading UI resource: ${tool.uiResourceUri}")

                    try {
                        // Read the UI resource
                        val request = ReadResourceRequest(ReadResourceRequestParams(uri = tool.uiResourceUri))
                        val resourceResult = client.readResource(request)
                        val htmlContent = resourceResult.contents.firstOrNull()?.let { content ->
                            when (content) {
                                is TextResourceContents -> content.text
                                is BlobResourceContents -> {
                                    String(android.util.Base64.decode(content.blob, android.util.Base64.DEFAULT))
                                }
                                else -> null
                            }
                        }

                        if (htmlContent != null) {
                            Log.i(TAG, "Loaded UI resource (${htmlContent.length} chars)")
                            // Store both HTML and tool result for AppBridge
                            val toolResultJson = json.encodeToString(
                                kotlinx.serialization.json.JsonObject.serializer(),
                                buildJsonObject {
                                    put("content", buildJsonArray {
                                        callResult.content.forEach { block ->
                                            // Extract text properly from content block
                                            val text = when (block) {
                                                is TextContent -> block.text
                                                else -> block.toString()
                                            }
                                            add(buildJsonObject {
                                                put("type", JsonPrimitive("text"))
                                                put("text", JsonPrimitive(text))
                                            })
                                        }
                                    })
                                    put("isError", JsonPrimitive(callResult.isError ?: false))
                                }
                            )
                            updateToolCall(toolCall.id) { it.copy(
                                state = ToolCallState.State.READY,
                                htmlContent = htmlContent,
                                toolResult = toolResultJson,
                                inputArgs = inputArgs?.let { args ->
                                    args.mapValues { (_, v) -> v.toString() }
                                }
                            )}
                        } else {
                            Log.w(TAG, "No HTML content in resource")
                            val resultText = callResult.content.joinToString("\n") { it.toString() }
                            updateToolCall(toolCall.id) { it.copy(
                                state = ToolCallState.State.COMPLETED,
                                result = resultText
                            )}
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to read UI resource: ${e.message}", e)
                        val resultText = callResult.content.joinToString("\n") { it.toString() }
                        updateToolCall(toolCall.id) { it.copy(
                            state = ToolCallState.State.COMPLETED,
                            result = resultText
                        )}
                    }
                } else {
                    // No UI resource, show text result
                    val resultText = callResult.content.joinToString("\n") { it.toString() }
                    Log.i(TAG, "Tool result (no UI): $resultText")
                    updateToolCall(toolCall.id) { it.copy(
                        state = ToolCallState.State.COMPLETED,
                        result = resultText
                    )}
                }

            } catch (e: Exception) {
                Log.e(TAG, "Tool call failed", e)
                updateToolCall(toolCall.id) { it.copy(
                    state = ToolCallState.State.ERROR,
                    error = e.message
                )}
            }
        }
    }

    /**
     * Request to close a tool call. For apps, this marks as destroying and waits
     * for teardown. For non-app results, removes immediately.
     */
    fun requestClose(toolCall: ToolCallState) {
        if (toolCall.hasApp) {
            // Mark as destroying - the WebView will send teardown and call completeClose
            updateToolCall(toolCall.id) { it.copy(isDestroying = true) }
        } else {
            // Non-app results close immediately
            completeClose(toolCall.id)
        }
    }

    /**
     * Complete the close after teardown response (or immediately for non-apps).
     */
    fun completeClose(id: String) {
        _toolCalls.value = _toolCalls.value.filter { it.id != id }
    }

    /**
     * Forward a tool call from the WebView App to the MCP server.
     * Returns the result as a JSON string.
     */
    suspend fun forwardToolCall(name: String, arguments: Map<String, Any>?): String {
        val client = mcpClient ?: throw IllegalStateException("Not connected")

        Log.i(TAG, "Forwarding tool call: $name with args: $arguments")

        val callResult = client.callTool(name, emptyMap(), emptyMap())

        // Format result as JSON for the App
        val resultJson = json.encodeToString(
            kotlinx.serialization.json.JsonObject.serializer(),
            buildJsonObject {
                put("content", buildJsonArray {
                    callResult.content.forEach { block ->
                        val text = when (block) {
                            is TextContent -> block.text
                            else -> block.toString()
                        }
                        add(buildJsonObject {
                            put("type", JsonPrimitive("text"))
                            put("text", JsonPrimitive(text))
                        })
                    }
                })
                put("isError", JsonPrimitive(callResult.isError ?: false))
            }
        )

        Log.i(TAG, "Tool call result: $resultJson")
        return resultJson
    }

    private fun updateToolCall(id: String, update: (ToolCallState) -> ToolCallState) {
        _toolCalls.value = _toolCalls.value.map { if (it.id == id) update(it) else it }
    }

    private fun generateDefaultInput(tool: ToolInfo): String {
        // TODO: Parse inputSchema and generate defaults
        return "{}"
    }
}
